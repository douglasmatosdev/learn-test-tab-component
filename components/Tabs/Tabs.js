import { useState, useEffect } from 'react';
import { useRouter } from 'next/router';
import styles from './Tabs.module.css';

import { slugfy } from '../../utils/slugfy';

const Tabs = ({ children, initialTab }) => {
    const [activeTab, setActiveTab] = useState(children[0].props.label);
    const router = useRouter();

    const handleClick = (e, newActiveTab) => {
        e.preventDefault();

        setActiveTab(slugfy(newActiveTab));
    }

    useEffect(() => {
        if (initialTab.tab) {
            setActiveTab(initialTab.tab);
            console.log(initialTab)
        }
    }, [])

    useEffect(() => {
        router.push(
            `${router.pathname}?tab=${slugfy(activeTab)}`,
            undefined,
            { shallow: true }
        );
        console.log(activeTab)
    }, [activeTab])

    return (
        <div>
            <ul className={styles.tabs}>
                {children.map((tab, i) => {
                    const label = tab.props.label;
                    return (
                        <li
                            key={i * 2}
                            className={slugfy(label) == activeTab ? styles.current : ''}
                        >
                            <a onClick={e => handleClick(e, label)} href="#">{label}</a>
                        </li>
                    )
                })}
            </ul>
            {children.map((one, i) => {
                if (slugfy(one.props.label) == activeTab) {
                    return (
                        <div key={i} className={styles.content}>
                            {one}
                        </div>
                    )
                }
            })}
        </div>
    )
}

export { Tabs }
