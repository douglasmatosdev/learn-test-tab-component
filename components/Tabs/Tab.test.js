import React from "react";
import ReactDOM from "react";
import { render } from "@testing-library/react";
import { Tabs } from "./Tabs";
import { useRouter } from "next/router";

jest.mock('next/router');

beforeEach(() => {
    useRouter.mockReturnValue({
        push: () => { },
    });
});

// TEST === === === === === === === === === === === === === === ===
test("<Tabs /> matches snapshot", () => {
    const component = render(
        <Tabs initialTab={{ tab: "tab-3" }}>
            <div label="Tab 1">Something</div>
            <div label="Tab 2">Else</div>
            <div label="Tab 3">Hello</div>
        </Tabs>
    );

    expect(component.container).toMatchSnapshot();
});

// TEST === === === === === === === === === === === === === === ===
test("<Tabs /> renders without crashing", () => {
    const div = document.createElement("div");

    render(
        <Tabs initialTab={{ tab: "tab-3" }}>
            <div label="Tab 1">Something</div>
            <div label="Tab 2">Else</div>
            <div label="Tab 3">Hello</div>
        </Tabs>,
        div
    );
});

// TEST === === === === === === === === === === === === === === ===
test("Defaults to the first tab's content", () => {
    const div = document.createElement("div");

    const { getByTestId } = render(
        <Tabs>
            <div label="Tab 1">Something</div>
            <div label="Tab 2">Else</div>
            <div label="Tab 3">Hello</div>
        </Tabs>,
        div
    );

    const content = getByTestId("content");
    expect(content.textContent).toBe("Something");
});

// TEST === === === === === === === === === === === === === === ===
test("Can set a different tab as the initial state", () => {
    const div = document.createElement("div");

    const { getByTestId } = render(
        <Tabs initialTab={{ tab: "tab-3" }}>
            <div label="Tab 1">Something</div>
            <div label="Tab 2">Else</div>
            <div label="Tab 3">Hello</div>
        </Tabs>,
        div
    );

    const content = getByTestId("content");
    expect(content.textContent).toBe("Hello");
});

// TEST === === === === === === === === === === === === === === ===
test("Click from one tab to be the next", () => {
    const div = document.createElement("div");

    const { getByTestId } = render(
        <Tabs initialTab={{ tab: "tab-3" }}>
            <div label="Tab 1">Something</div>
            <div label="Tab 2">Else</div>
            <div label="Tab 3">Hello</div>
        </Tabs>,
        div
    );

    const content = getByTestId("content");
    expect(content.textContent).toBe("Hello");

    // diferentiate the current tab

    // click on a new tab

    // check the new content
});