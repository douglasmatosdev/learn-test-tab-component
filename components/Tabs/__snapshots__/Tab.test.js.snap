// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> matches snapshot 1`] = `
<div>
  <div>
    <ul
      class="tabs"
    >
      <li
        class=""
      >
        <a
          href="#"
        >
          Tab 1
        </a>
      </li>
      <li
        class=""
      >
        <a
          href="#"
        >
          Tab 2
        </a>
      </li>
      <li
        class="current"
      >
        <a
          href="#"
        >
          Tab 3
        </a>
      </li>
    </ul>
    <div
      class="content"
      data-testid="content"
    >
      <div
        label="Tab 3"
      >
        Hello
      </div>
    </div>
  </div>
</div>
`;
